### DISCUSSION TITLE
Docker Compose libcups.so.2

### DISCUSSION THREAD

**Discussion by danielchild1** [OpenPrinting/cups#1012] [Q&A]
When I use `docker compose up` I get caught on the the error ` error while loading shared libraries: libcups.so.2: cannot open shared object file: No such file or directory`. So I go into the [Dockerfile](https://github.com/OpenPrinting/cups/blob/8ed4737896b04617383970fbb0bed96a2825edd8/Dockerfile) and add `libcups2` to the end of the [apt-get install -y](https://github.com/OpenPrinting/cups/blob/8ed4737896b04617383970fbb0bed96a2825edd8/Dockerfile#L13) line. Once I rebuild and run it gives me a `symbol lookup error: /usr/sbin/cupsd: undefined symbol: _cupsArrayStrcmp` Error.   <br>
I am wondering if I am needing to do some more prepwork before running this docker container. Is there documentation for the docker-compose.yml? I cannot find it. Thank you!

**Comment by coolstuff99** [comment]
Hey @danielchild1 I am having same issue... did you find way to solve it?

**Comment by coolstuff99** [comment]
@michaelrsweet by any chance can you bring us some light here? Thank you so much.

**Reply by michaelrsweet** [reply]
The current Github master code *should* work better. Linux needs the "ldconfig" command to be run, otherwise the cache of symbols in libcups.so.2 isn't updated, and that can prevent programs using the newer functions from working. I also cleaned up the Docker stuff so that it would build properly (although a fully-functional Docker image will really depend on more of the OpenPrinting components...)

**Reply by anony253** [reply]
I also have this problem, can you please explain in a bit more detail how is this exactly intended to be used, as I have no clue where to run the "ldconfig" command, on the host, or in the container? If in the container, at which point, since it is impossible to even start this container in the first place. I have tried both in Dockerfile and in docker-compose.yaml but with no success

**Reply by michaelrsweet** [reply]
@anony253 Can you provide more information? What errors are you seeing when you try building and running the container?

**Reply by anony253** [reply]
Exactly what the original ticket was for:
`error while loading shared libraries: libcups.so.2: cannot open shared object file: No such file or directory`

**Comment by Magniveo** [comment]
May be will help, change pathtolib where you install cups.
```
#!/bin/bash
#set -x
# Set environment variables
export CUPS_SERVERBIN="/home/pathtolib/usr/local/lib64/cups"
export LD_LIBRARY_PATH="/home/pathtolib/usr/local/lib64:$LD_LIBRARY_PATH"
export CUPS_DATADIR="/home/pathtolib/usr/local/share/cups"
export CUPS_REQUESTROOT="/home/pathtolib/var/run/cups"
export CUPS_STATEDIR="/home/pathtolib/var/run/cups"
export CUPS_FONTPATH="/home/pathtolib/usr/local/share/cups/fonts"


# Test configuration file
/home/pathtolib/usr/local/sbin/cupsd -t -c /home/pathtolib/usr/local/etc/cups/cupsd.conf
if [ $? -ne 0 ]; then
  echo "Configuration file test failed. Exiting."
  exit 1
fi

# Run cupsd in foreground mode for debugging
echo "Starting cupsd..."
# Run cupsd with a custom config file
/home/pathtolib/usr/local/sbin/cupsd -f -l -c /home/pathtolib/usr/local/etc/cups/cupsd.conf
```

**Comment by SeanTasker** [comment]
I ran into this issue be the case on both `2.4.x` and `master`. Which I later realised were the same revision. The instructions below can help you fix the problem but I also ran into further issues with an assertion `cupsd: browser.c:245: avahi_domain_browser_free: Assertion `b' failed.` (I had to drop into an interactive shell then run `/usr/sbin/cupsd -f` to see this error. If below is all that was required to fix this issue I would have been happy to open a pull request.

I had more luck running [olbat/cupsd](https://hub.docker.com/r/olbat/cupsd) instead of building from source.

```bash
docker run -d -p 631:631 -v /var/run/dbus:/var/run/dbus --name cupsd olbat/cupsd
```

### Fixes
There are two fixes that should resolve libraries not being found. Modify `Makedefs.in` to make `ALL_LDFLAGS` include the flags for LDFLAGS or modify `Dockerfile`.

### Dockerfile change
```diff
-RUN ./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var && make clean && make && make install
+RUN LDFLAGS='-Wl,-rpath=${libdir}' ./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var && make clean && make && make install
```

### Makedefs.in version
```diff
-ALL_LDFLAGS    =       -L../cups @LDARCHFLAGS@ @RELROFLAGS@ $(LDFLAGS)  \
+ALL_LDFLAGS    =       -L../cups -Wl,-rpath=${libdir} @LDARCHFLAGS@ @RELROFLAGS@ $(LDFLAGS)  \
```
The Dockerfile change should source rebuild when you when you run `docker build .` again but the second change may require cleaning your docker cache.

### To Clean

**This assumes you only have one image with "cups-cups" in the name**
```bash
# Remove containers referencing the cups image
docker container ls -a | grep 'cups-cups' | awk '{print $1}' | xargs -I{} docker container rm "{}"
# Remove cups image
docker image rm cups-cups
docker system prune
```
Otherwise run `docker container ls -a` and then `docker container rm containerID` for each container that references the cups image before pruning. Run `docker build .` when you've made the change required.
